sbi/inference/snpe/snpe_c.py

```diff
@@ -231,5 +370,354 @@ def _log_prob_proposal_posterior(
+
+        return precisions_pp, covariances_pp
+
+    def _means_proposal_posterior(
+        self,
+        covariances_pp: Tensor,
+        means_p: Tensor,
+        precisions_p: Tensor,
+        means_d: Tensor,
+        precisions_d: Tensor,
+    ):
+        """
+        Return the means of the proposal posterior.
+
+        means_pp = C_ix * (P_i * m_i + P_x * m_x - P_o * m_o).
+
+        Args:
+            covariances_pp: Covariance matrices of the proposal posterior.
+            means_p: Means of the proposal distribution.
+            precisions_p: Precision matrices of the proposal distribution.
+            means_d: Means of the density estimator.
+            precisions_d: Precision matrices of the density estimator.
+
+        Returns: Means of the proposal posterior. L*K terms.
+        """
+
+        num_comps_p = precisions_p.shape[1]
+        num_comps_d = precisions_d.shape[1]
+
+        # First, compute the product P_i * m_i and P_j * m_j
+        cov_m_prod_p = _batched_mv(precisions_p, means_p)
+        cov_m_prod_d = _batched_mv(precisions_d, means_d)
+
+        # Repeat them to allow for matrix operations: same trick as for the precisions.
+        cov_m_prod_p_rep = cov_m_prod_p.repeat_interleave(num_comps_d, dim=1)
+        cov_m_prod_d_rep = cov_m_prod_d.repeat(1, num_comps_p, 1)
+
+        # Means = C_ij * (P_i * m_i + P_j * m_j).
+        summed_cov_m_prod_rep = cov_m_prod_p_rep + cov_m_prod_d_rep
+        if isinstance(self._maybe_z_scored_prior, mvn.MultivariateNormal):
+            summed_cov_m_prod_rep -= self.prec_m_prod_prior
+
+        means_pp = _batched_mv(covariances_pp, summed_cov_m_prod_rep)
+
+        return means_pp
+
+    @staticmethod
+    def _logits_proposal_posterior(
+        means_pp: Tensor,
+        precisions_pp: Tensor,
+        covariances_pp: Tensor,
+        logits_p: Tensor,
+        means_p: Tensor,
+        precisions_p: Tensor,
+        logits_d: Tensor,
+        means_d: Tensor,
+        precisions_d: Tensor,
+    ):
+        """
+        Return the component weights (i.e. logits) of the proposal posterior.
+
+        Args:
+            means_pp: Means of the proposal posterior.
+            precisions_pp: Precision matrices of the proposal posterior.
+            covariances_pp: Covariance matrices of the proposal posterior.
+            logits_p: Component weights (i.e. logits) of the proposal distribution.
+            means_p: Means of the proposal distribution.
+            precisions_p: Precision matrices of the proposal distribution.
+            logits_d: Component weights (i.e. logits) of the density estimator.
+            means_d: Means of the density estimator.
+            precisions_d: Precision matrices of the density estimator.
+
+        Returns: Component weights of the proposal posterior. L*K terms.
+        """
+
+        num_comps_p = precisions_p.shape[1]
+        num_comps_d = precisions_d.shape[1]
+
+        # Compute alpha_i * beta_j
```

Comment:
this will be the log logits, no?

Reply:
logits are always log(). (There had been some inconsistencies in the code with that, but they are fixed now. So, now, logit = log(component_weight).

But yes, I fixed the comment ;)